File: Task_1/task_1.cpp
#include <iostream>
#include <vector>

using namespace std;

// Функция для ввода данных и вычислений
int getRiverCrossingResult() {
    int stones, n;

    cout << "Введите общее количество валунов (stones): ";
    cin >> stones; // Читаем количество валунов с консоли

    cout << "Введите количество птиц (N): ";
    cin >> n;     // Читаем количество птиц с консоли

    vector<int> birds(n); // Создаем вектор для хранения дальности полета птиц
    cout << "Введите дальность полета для каждой из " << n << " птиц через пробел: ";
    for (int i = 0; i < n; ++i) {
        cin >> birds[i]; // Читаем дальность полета каждой птицы
    }

    vector<bool> visitedStones(stones, false); // Создаем вектор для отслеживания посещенных валунов.
                                             // Изначально все значения false (валун не посещен)

    // Моделируем перелеты каждой птицы
    for (int range : birds) {
        for (int i = range - 1; i < stones; i += range) {
            visitedStones[i] = true; // Помечаем валун как посещенный
        }
    }

    int unvisitedCount = 0;
    for (bool visited : visitedStones) {
        if (!visited) {        // Если валун не был посещен
            unvisitedCount++; // увеличиваем счетчик
        }
    }

    return unvisitedCount; // Возвращаем количество непосещенных валунов
}

int main() {
    int result = getRiverCrossingResult(); // Вызываем функцию для получения результата
    cout << "Количество валунов, на которых не появится ни одной птицы: " << result << endl; // Выводим результат
    return 0;
}

File: Task_1/task_1.php
<?php
function getRiverCrossingResult() {
    echo "Введите общее количество валунов (stones): ";
    $stones = intval(readline()); // Читаем количество валунов

    echo "Введите количество птиц (N): ";
    $n = intval(readline());     // Читаем количество птиц

    echo "Введите дальность полета для каждой из $n птиц через пробел: ";
    $birdDistances = explode(" ", readline()); // Читаем дальности полета и разбиваем строку в массив
    $birds = array_map('intval', $birdDistances); // Преобразуем строковые значения в числа

    $visitedStones = array_fill(0, $stones, false); // Создаем массив для отслеживания посещенных валунов.
                                                    // Изначально все значения false
    // Моделируем перелеты птиц
    foreach ($birds as $range) {
        for ($i = $range - 1; $i < $stones; $i += $range) {
            $visitedStones[$i] = true; // Помечаем валун как посещенный
        }
    }

    $unvisitedCount = 0;
    foreach ($visitedStones as $visited) {
        if (!$visited) {         // Если валун не посещен
            $unvisitedCount++; // Увеличиваем счетчик
        }
    }

    return $unvisitedCount; // Возвращаем количество непосещенных валунов
}

$result = getRiverCrossingResult(); // Вызываем функцию
echo "Количество валунов, на которых не появится ни одной птицы: " . $result . PHP_EOL; // Выводим результат
?>
